File: rosalind/laff/rosalind_laff.cpp
/*
============================================================
http://rosalind.info/problems/laff

Given: Two protein strings s and t in FASTA format (each
having length at most 10,000 aa).

Return: The maximum local alignment score of s and t,
followed by substrings r and u of s and t, respectively,
that correspond to the optimal local alignment of s and t

Use:
The BLOSUM62 scoring matrix.
Gap opening penalty equal to 11.
Gap extension penalty equal to 1.

If multiple solutions exist, then you may output any one.
============================================================
*/
#include <algorithm>
#include <sstream>
#include <iostream>
#include <fstream>
#include <climits>
#include <ctime>
#include <string>
#include <vector>
#include <map>
using namespace std;

#define MINUS_INFINITY -1000000
#define NONE -1

class Score {
public:
  map< char , int > index;
  vector< vector< int > > m;

  Score(const string& file_name)
  {
    ifstream infile(file_name.c_str());
    string line;
    int count = -1;
    cout << "Reading " << file_name << "\n";
    while (getline(infile, line))
      {
	//	cout << line << "\n";
	istringstream iss(line);
	if (count == -1) {
	  char key;
	  vector<char> keys;
	  while (iss >> key) keys.push_back(key);
 	  int n = keys.size();
	  //	  for (int i = 0; i < n; i++) cout << keys[i] << " ";
	  //	  cout << "\n";
	  m.resize(n);
	  for (int i = 0; i < n; i++) {
	    index[keys[i]] = i;
	    m[i].resize(n);
	  }
	} else {
	  char key;
	  iss >> key;
	  map< char, int > row;
	  int value;
	  int j = -1;
	  while (iss >> value) {
	    j++;
	    m[count][j] = value;
	  }
	}
	count++;
      }
  }
};

/*void alignment_matrix_form(const string& x, const string& y, int** c, int** I, int** J, int i, int j, 
			   const string& gap_symbol)
{
}*/

int ** createIntMatrix(unsigned int rows, unsigned int cols) {
  // Initialize an int matrix with zeros
  cout << "Allocating matrix" << rows << "x" << cols << "\n";
  int ** matrix;
  matrix = (int **) calloc(rows, sizeof(int *));
  for(unsigned int i = 0; i < rows; i++)
    matrix[i] = (int *) calloc(cols, sizeof(int));
  return matrix;
}

void freeIntMatrix(int** matrix, unsigned int rows, unsigned int cols) {
  for(unsigned int i = 0; i < rows; i++)
    free(matrix[i]);
  free(matrix);
}

void optimal_alignment_matrix_form(string& x, string& y, 
				  Score& score, int gap_init, int gap_ext,
				  int debug)
{
  /*
    Optimal local alignment that maximizes the alignment score based on the (symmetric) score score
    and, initial gap penalty gap_init < 0, and a gap extension penalty gap_ext < 0.
    Integrated DP+backtracking, O(mn) time+storage.
  */
  const int RESTART = 0;
  const int m = x.size();
  const int n = y.size();
  clock_t start = clock();
  if (debug >= 1)
    cout << m << " " << n << "\n";
  // All arrays initialized to 0
  int** c = createIntMatrix(m+1, n+1);
  int** d = createIntMatrix(m+1, n+1);
  int** e = createIntMatrix(m+1, n+1);
  int** I = createIntMatrix(m+1, n+1);
  int** J = createIntMatrix(m+1, n+1);
  for (int i = 0; i <= m; i++) {
    for (int j = 0; j <= n; j++) {
      c[i][j] = 0;
      d[i][j] = 0;
      e[i][j] = 0;
      I[i][j] = 0;
      J[i][j] = 0;
    }
  }

  //-----------------------------------
  // Initial condition, first row
  //-----------------------------------
  if (debug >= 1)
    cout << "Init first row\n";
  for (int j = 0; j <= n; j++) d[0][j] = MINUS_INFINITY;
  // Best: empty substring alignment (score=0)
  for (int j = 0; j <= n; j++) {
    e[0][j] = 0;
    I[0][j] = NONE;
    J[0][j] = NONE;
   }
   
  //-----------------------------------
  // Initial condition, first column
  //-----------------------------------
  for (int i = 0; i <= m; i++) d[i][0] = MINUS_INFINITY;
  // Best: empty substring alignment (score=0)
  for (int i = 0; i <= m; i++) {
    e[i][0] = 0;
    I[i][0] = NONE;
    J[i][0] = NONE;
  }
 
  //-----------------------------------
  // Dynamic programming
  //-----------------------------------
  for (int i = 1; i <= m; i++) {
    //    cout << i << "\n";
    char xi = x[i-1];
    if (debug >= 1) {
      if (i % 100 == 0) {
	clock_t t = clock();
	double elapsed_secs = double(t - start) / CLOCKS_PER_SEC;
	cout << i << " " << elapsed_secs << "\n";
      }
    }
    for (int j = 1; j <= n; j++) {
      char yj = y[j-1];
      // Possible types of paths to extend to (i,j)
      //      cout << xi << " " << yj << " " << score.m[score.index[xi]][score.index[yj]] << "\n";
      int d_max = c[i - 1][j - 1] + score.m[score.index[xi]][score.index[yj]];
      int e1 = d[i - 1][j] + gap_init;
      int e2 = e[i - 1][j] + gap_ext;
      int e3 = d[i][j - 1] + gap_init;
      int e4 = e[i][j - 1] + gap_ext;
      int e_max = max(max(max(e1, e2), e3), e4);
      int c_max = max(max(d_max, e_max), RESTART);  // Pick best path (greedy approach)
      // Reconstruct the previous element along the best path             
      int ip, jp;
      if (c_max == d_max) { ip = i - 1; jp = j - 1; }
      else if (c_max == e_max) { 
	if ((e_max == e1) || (e_max == e2)) { ip = i - 1; jp = j; }
	else { ip = i; jp = j - 1; }
      } else { ip = NONE; jp = NONE; }
      // Save (i,j) state in matrix
      d[i][j] = d_max;
      e[i][j] = e_max;
      c[i][j] = c_max;
      I[i][j] = ip;
      J[i][j] = jp;
    }
  }
    
  // Maximize score, back-track path
  // Find argmax(c)
  int c_max = c[0][0], i_max, j_max;
  for (int i = 0; i <= m; i++) {
    for (int j = 0; j <= n; j++) {
      int cij = c[i][j];
      if (cij > c_max) {
	c_max = cij;
	i_max = i;
	j_max = j;
      }
    }
  }
  /*
  for (int i = 0; i <= m; i++) {
    for (int j = 0; j <= n; j++)
      cout << c[i][j] << " ";
    cout << "\n";
  }
  */
  
  //  if (debug >= 1) cout << "Starting back-tracking at " << i_max << " " << j_max << "\n";
  if (debug >= 2) {
    cout << c;
    cout << I;
    cout << J;
  }
  cout << c[i_max][j_max] << "\n";

  /*
    Return the augmented-aligned strings of the original strings x and y
    from the DP matrix (result of global_alignment_matrix()) c.
  */
  string gap_symbol("-");
  string s = "";
  string t = "";
  int i = i_max;
  int j = j_max;
  while (true) {
    //    cout << i << " " << j << "\n";
    int ip = I[i][j];
    int jp = J[i][j];
    if ((ip == NONE) || (jp == NONE)) break;
    if (ip == i) s = gap_symbol + s;
    else s = x[i-1] + s;
    if (jp == j) t = gap_symbol + t;
    else t = y[j-1] + t;
    i = ip;
    j = jp;
  }
  cout << s << "\n";
  cout << t << "\n";

  freeIntMatrix(c, m+1, n+1);
  freeIntMatrix(d, m+1, n+1);
  freeIntMatrix(e, m+1, n+1);
  freeIntMatrix(I, m+1, n+1);
  freeIntMatrix(J, m+1, n+1);
  //  alignment_matrix_form(x, y, c, I, J, i_max, j_max, gap_symbol);
}

vector<string> fafsa_values(const string& file_name)
{
  vector<string> s;
  ifstream infile(file_name.c_str());
  string line;
  int count = -1;
  while (getline(infile, line))
    {
      //      line[line.size()-1] = '\0';
      //      line.erase(remove(line.begin(), line.end(), '\n'), line.end());
      //      cout << "'" << line << "'\n";
      if (line[0] == '>') {
	count++;
	string v("");
	s.push_back(v);
	cout << count << "\n";
      } else {
	s[count].append(line.substr(0, line.size()-1));
      }
    }
  return s;
}
    
void laff(string f, int debug)
{
  /* Main driver to solve the LAFF problem. */
  vector<string> s = fafsa_values(f);
  for (int i = 0; i < s.size(); i++) cout << "'" << s[i] << "'" << "\n";
  Score score("c:/ober/code/misc/rosalind/blosum62.dat");
  optimal_alignment_matrix_form(s[0], s[1], score, -11, -1, debug);
}

int main(int argc, char **argv) {
  laff(argv[1], 1);
  //  laff("rosalind_laff_sample2.dat", 1);
  //  laff("rosalind_laff.dat", 1);
  return 0;
}

File: rosalind/laff/rosalind_laff.py
'''
============================================================
http://rosalind.info/problems/laff

A local alignment of two strings s and t is an alignment of
substrings r and u of s and t, respectively. Let opt(r,u)
denote the score of an optimal alignment of r and u with
respect to some predetermined alignment score.

Given: Two protein strings s and t in FASTA format (each
having length at most 1000 aa).

Return: A maximum alignment score along with substrings r
and u of s and t, respectively, which produce this maximum
alignment score (multiple solutions may exist, in which case
you may output any one). Use:

The PAM250 scoring matrix.
Linear gap penalty equal to 5.
============================================================
'''
import rosalind.rosutil as ro, rosalind.rosalign as ra, numpy as np, time, sys

'''Minus infinity = smallest possible (integer) cost. A little larger than min integer so that we don\'t
wrap around when subtracting gap penalties.'''
MINUS_INFINITY = -sys.maxint - 1 + 10000

'''Index of alignment path beginning.'''  
NONE = -1

def optimal_alignment_array_form(x, y, score, gap_init, gap_ext, debug=0, gap_symbol='-'):
    '''Optimal alignment that maximizes the alignment score based on the (symmetric) ScoringMatrix score
    and, initial gap penalty \'gap_init\' < 0, and a gap extension penalty \'gap_ext\' < 0.
    If align='local', performs local alignment, otherwise global.
    Integrated DP+backtracking, O(mn) time+storage, but may be faster (separate score,path arrays).'''
    RESTART = 0
    m, n = len(x), len(y)
    if debug >= 1:
        print m, n
        start = time.time()
    c = [[0 for _ in xrange(n + 1)] for _ in xrange(m + 1)]  # Max score of paths ending at (i,j)
    d = [[0 for _ in xrange(n + 1)] for _ in xrange(m + 1)]  # Max score of paths ending at (i,j), ending with a substitution
    e = [[0 for _ in xrange(n + 1)] for _ in xrange(m + 1)]  # Max score of paths ending at (i,j), ending with an ins/del
    p = [[0 for _ in xrange(n + 1)] for _ in xrange(m + 1)]
    
    #-----------------------------------
    # Initial condition, first row & col
    #-----------------------------------
    # Best: empty substring alignment (score=0)
    for j in xrange(n + 1):
        d[0][j] = MINUS_INFINITY
        e[0][j] = 0 
        p[0][j] = NONE
    # Best: empty substring alignment (score=0)
    for i in xrange(m + 1):
        d[i][0] = MINUS_INFINITY
        e[i][0] = 0 
        p[i][0] = NONE
    
    #-----------------------------------
    # Dynamic programming
    #-----------------------------------
    for i, xi in enumerate(x, 1):
        if debug:
            if i % 100 == 0: print i, time.time() - start
        for j, yj in enumerate(y, 1):
            i1, j1 = i - 1, j - 1
            # Possible types of paths to extend to (i,j)
            d[i][j] = d_max = c[i1][j1] + score(xi, yj)
            e1, e2, e3, e4 = \
            d[i1][j] + gap_init, e[i1][j] + gap_ext, \
            d[i][j1] + gap_init, e[i][j1] + gap_ext
            e[i][j] = e_max = max(e1, e2, e3, e4)
            c[i][j] = c_max = max(d_max, e_max, RESTART)  # Pick best path (greedy approach)
            # Reconstruct the previous element along the best path             
            if c_max == d_max: p_max = ra.EQ
            elif c_max == e_max: p_max = ra.T_GAP if e_max == e1 or e_max == e2 else ra.S_GAP 
            else: p_max = NONE
            # Save (i,j) state in matrix
            p[i][j] = p_max
        
    # Maximize score, back-track path
    i, j = divmod(np.argmax(c), n + 1)
    if debug >= 1: print 'Starting back-tracking at', (i, j)
    if debug >= 2:
        print np.array(c)
        print np.array(p)
    return c[i][j], alignment_array_form(x, y, c, p, (i, j), gap_symbol=gap_symbol)

def alignment_array_form(x, y, c, p, (i, j), gap_symbol='-'):
    '''Return the augmented-aligned strings of the original strings x and y
    from the DP matrix (result of global_alignment_matrix()) c.'''
    s, t, ptr = '', '', ra.EQ
    while ptr != NONE:
        ptr = p[i][j]
        # print i, j, ptr, s, t
        if ptr == ra.EQ: i, j, si, ti = i - 1, j - 1 , x[i - 1], y[j - 1]
        elif ptr == ra.T_GAP: i, si, ti = i - 1, x[i - 1], gap_symbol
        else: j, si, ti = j - 1, gap_symbol, y[j - 1]
        s, t = si + s, ti + t
    return s, t

def laff(f, debug=0):
    '''Main driver to solve the LOCA problem.'''
    x, y = ro.fafsa_values(f)
    d, (xx, yy) = optimal_alignment_array_form(x, y, ra.BLOSUM62, -11, -1, debug=debug, gap_symbol='-')
    if debug:
        print x
        print y 
        print ra.score_of(xx, yy, ra.BLOSUM62, -11, -1)  # , debug=True)
    return ro.join_list((d, xx, yy), delimiter='\n')

from itertools import product

def laff_yesimon(f):
    score = ra.BLOSUM62
    s, t = ro.fafsa_values(f)
    sl, tl = len(s), len(t)
    T = np.zeros((sl + 1, tl + 1), dtype=np.character)
    V = np.zeros((sl + 1, tl + 1))
    F = np.zeros((sl + 1, tl + 1))
    G = np.zeros((sl + 1, tl + 1))
    for i, j in product(xrange(1, sl + 1), xrange(1, tl + 1)):
        cost = score(s[i - 1], t[j - 1])
        F[i, j] = V[i - 1, j - 1] + cost
        G[i, j] = max(F[i - 1, j] - 12, G[i - 1, j] - 1, F[i, j - 1] - 12, G[i, j - 1] - 1)
        V[i, j] = v = max(F[i, j], G[i, j], 0)
        if v == F[i, j]: T[i, j] = 'D'
        elif v == G[i, j]:
            if v == F[i - 1, j] - 12 or G[i - 1, j] - 1: T[i, j] = 'L'
            else: T[i, j] = 'U'
        elif v == 0: T[i, j] = ''
    i, j = np.unravel_index(np.argmax(V), V.shape)
    print(int(V[i, j]))
    sa = ''
    ta = ''
    while T[i, j]:
        direction = T[i, j]
        if direction == 'D':
            sa += s[i - 1]
            ta += t[j - 1]
            i, j = i - 1, j - 1
        elif direction == 'L':
            ta += t[j - 1]
            i = i - 1
        elif direction == 'U':
            sa += s[i - 1]
            j = j - 1
    print(''.join(reversed(sa)))
    print(''.join(reversed(ta)))
    
if __name__ == "__main__":
    # print ra.optimal_alignment('PLS', 'MN', ra.BLOSUM62, -11, -1, debug=True)
    # print ra.score_of('LS', 'MN', ra.BLOSUM62, -11, -1, debug=True)
#     print ra.score_of('TLY', 'NLY', ra.BLOSUM62, -11, -1, debug=True)
# 
#     print ra.optimal_alignment('LEAS', 'MEAN', ra.BLOSUM62, -11, -1, debug=True)
    # print laff('rosalind_laff_sample.dat')
#  print laff('rosalind_laff.dat', debug=1)

#     cProfile.run('a()', 'a')
#     p = pstats.Stats('a').strip_dirs()
#     p.sort_stats('cumulative').print_stats(50)
    print laff('rosalind_laff_sample4.dat', debug=2)
    print laff_yesimon('rosalind_laff_sample4.dat')

#     print laff('rosalind_laff_sample5.dat', debug=1)
#     print ra.score_of('SAVAKC------HQR', 'SAKQKCARIENSHQR', ra.BLOSUM62, -11, -1, debug=True)
    
    print laff('rosalind_laff_sample3.dat', debug=2)
    print laff_yesimon('rosalind_laff_sample3.dat')
